sdk/src/transaction_context.rs

```diff
@@ -555,6 +611,28 @@ impl<'a> BorrowedAccount<'a> {
 
     /// Assignes the owner of this account (transaction wide)
     pub fn set_owner(&mut self, pubkey: &[u8]) -> Result<(), InstructionError> {
+        if self
+            .transaction_context
+            .enable_early_verification_of_account_modifications
+        {
+            // Only the owner can assign a new owner
+            if !self.is_owned_by_current_program() {
+                return Err(InstructionError::ModifiedProgramId);
+            }
+            // and only if the account is writable
+            if !self.is_writable() {
+                return Err(InstructionError::ModifiedProgramId);
+            }
+            // and only if the account is not executable
+            if self.is_executable() {
+                return Err(InstructionError::ModifiedProgramId);
+            }
+            // and only if the data is zero-initialized or empty
+            if !is_zeroed(self.get_data()) {
+                return Err(InstructionError::ModifiedProgramId);
+            }
+        }
+        self.touch()?;
```

Comment:
these `.touch()` invocations are really need to be outside the feature flag? I think `self.touch()` is almost all cases infallible currently. but,  if they're not actually (in the future), this might risk cluster wide fork while the cluster is mixed with this new or original code, right?

it looks like the ultimate use is actually feature-gated here: https://github.com/solana-labs/solana/pull/25899/files#r916566935 (via not-feature gated new summation code in the `From`; this might be moved inside a feature gate maybe?)

Reply:
also, I'm wondering where failed account mutations should be counted towards to the summation or not. maybe i need precise semantics clarification (still digging relevant code).

i mean, data mutations which is prevented by the new checks won't be marked as `touch`-ed. So, failed ExecutionRecord in such a way won't include these attempts.

Reply:
@Lichtso I agree with Ryo's points here, can you take another look at this in case you lost track of these review comments?

Reply:
@ryoqun @jstarry We could move the `touch()` inside a `if new_value != old_value` up to the beginning of each setter. That way failed attempts would also count as touched.

However, the statistic would still be different. Because currently one can fail many attempts inside one Instruction, but after this feature the first failed attempt is always also the last one. So it implicitly limits the number of possible failures to one at maximum.

Reply:
I was agreeing with Ryo's first suggestion of moving `touch()` inside the feature gated code block. Specifically after the equality check.

I just looked closer into failed account mutations and agree that some of the stats shouldn't be updated (changed accounts and changed data size) but I think it's best to try to maintain the current behavior for now and address that later.